### DISCUSSION TITLE
BCR pre-processing

### DISCUSSION THREAD

**Discussion by saramoein372** [tuonglab/dandelion#114] [General]
Would you please help me to figure out the problem with BSR Pre_processing. For the step 4 I get this error:

ddl.pp.assign_isotypes(samples, blastdb = "/Users/saramoein/Downloads/dandelion-master/container/database/blast/human/human_BCR_C.fasta")

The error:

FileNotFoundError                         Traceback (most recent call last)
<ipython-input-14-bd4eef96649c> in <module>
----> 1 ddl.pp.assign_isotypes(samples, blastdb = "/Users/saramoein/Downloads/dandelion-master/container/database/blast/human/human_BCR_C.fasta")
      2 
      3 
      4 

/opt/anaconda3/lib/python3.8/site-packages/dandelion/preprocessing/_preprocessing.py in assign_isotypes(fastas, fileformat, org, correct_c_call, correction_dict, plot, save_plot, show_plot, figsize, blastdb, allele, parallel, ncpu, filename_prefix, verbose)
   1004 
   1005     for i in range(0, len(fastas)):
-> 1006         assign_isotype(fastas[i],
   1007                        fileformat=fileformat,
   1008                        org=org,

/opt/anaconda3/lib/python3.8/site-packages/dandelion/preprocessing/_preprocessing.py in assign_isotype(fasta, fileformat, org, correct_c_call, correction_dict, plot, save_plot, show_plot, figsize, blastdb, allele, parallel, ncpu, filename_prefix, verbose)
    818     if verbose:
    819         print('Running blastn \n')
--> 820     _run_blastn(filePath, blastdb, format_dict[fileformat], org, verbose)
    821     # parsing output into a summary.txt file
    822     if verbose:

/opt/anaconda3/lib/python3.8/site-packages/dandelion/preprocessing/_preprocessing.py in _run_blastn(fasta, blastdb, fileformat, org, verbose)
    416             print('Running command: %s\n' % (' '.join(cmd)))
    417         with open(blast_out, 'w') as out:
--> 418             run(cmd, stdout=out, env=env)
    419 
    420     def _parse_BLAST(fasta, fileformat):

/opt/anaconda3/lib/python3.8/subprocess.py in run(input, capture_output, timeout, check, *popenargs, **kwargs)
    491         kwargs['stderr'] = PIPE
    492 
--> 493     with Popen(*popenargs, **kwargs) as process:
    494         try:
    495             stdout, stderr = process.communicate(input, timeout=timeout)

/opt/anaconda3/lib/python3.8/subprocess.py in __init__(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, encoding, errors, text)
    856                             encoding=encoding, errors=errors)
    857 
--> 858             self._execute_child(args, executable, preexec_fn, close_fds,
    859                                 pass_fds, cwd, env,
    860                                 startupinfo, creationflags, shell,

/opt/anaconda3/lib/python3.8/subprocess.py in _execute_child(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, start_new_session)
   1704                     if errno_num != 0:
   1705                         err_msg = os.strerror(errno_num)
-> 1706                     raise child_exception_type(errno_num, err_msg, err_filename)
   1707                 raise child_exception_type(err_msg)
   1708 

FileNotFoundError: [Errno 2] No such file or directory: 'blastn'

Thanks

**Comment by zktuong** [comment]
Hi @saramoein372, thank you for interest in this package.

Your error is saying that it cannot find blastn in your path. blastn comes with blast+ (and not with igblast) so you might want to check if you have it with:
```bash
which blastn
```
It should return with an output like this:
```bash
/Users/kt16/Softwares/ncbi-blast-2.12.0+/bin/blastn
```

if not, download it (manually or with conda), and add it to your path:
```bash
export PATH=/path/to/folder/where/blastn/is:$PATH
```

To add it permanently, you can consider exporting it to your `~/.bashrc` or `~/.bash_profile` or equivalent files.
You can refer to the external software section of https://sc-dandelion.readthedocs.io/en/latest/README.html#installation.

Hope that helps!

**Comment by saramoein372** [comment]
Hi Kelvin,

Thank you for the help! I could solve the problem in step4.
I just have one question left for the step 3. When I run the step3, it give me this error:
```python
ddl.pp.reassign_alleles(samples[0:1], combined_folder = '/users/saramoein/downloads/BCR/tutorial_scgp2', germline = "/Users/saramoein/Downloads/dandelion-master/container/database/germlines/imgt/human/vdj")


Running tigger-genotype with novel allele discovery.
      Novel allele discovery execution halted.
      Attempting to run tigger-genotype without novel allele discovery.
     Insufficient contigs for running tigger-genotype. Defaulting to original heavy chain v_calls.
            Reconstructing heavy chain dmask germline sequences with v_call.
            Reconstructing light chain dmask germline sequences with v_call.
      For convenience, entries for heavy chain in `v_call` are copied to `v_call_genotyped`.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-16-21002daefa5a> in <module>
----> 1 ddl.pp.reassign_alleles(samples[0:1], combined_folder = '/users/saramoein/downloads/BCR/tutorial_scgp2', germline = "/Users/saramoein/Downloads/dandelion-master/container/database/germlines/imgt/human/vdj")
      2 
      3 
      4 
      5 

/opt/anaconda3/lib/python3.8/site-packages/dandelion/preprocessing/_preprocessing.py in reassign_alleles(data, combined_folder, v_germline, germline, org, v_field, germ_types, novel, cloned, plot, save_plot, show_plot, figsize, sample_id_dictionary, filename_prefix, verbose)
   1388             '      For convenience, entries for heavy chain in `v_call` are copied to `v_call_genotyped`.'
   1389         )
-> 1390         heavy = load_data(outDir + '/' + outDir + '_heavy' +
   1391                           germpass_dict[fileformat])
   1392         heavy['v_call_genotyped'] = heavy['v_call']

/opt/anaconda3/lib/python3.8/site-packages/dandelion/utilities/_core.py in load_data(obj)
   1035         obj_ = obj.copy()
   1036     else:
-> 1037         raise TypeError(
   1038             "Either input is not of <class 'pandas.core.frame.DataFrame'> or file does not exist."
   1039         )

TypeError: Either input is not of <class 'pandas.core.frame.DataFrame'> or file does not exist.
```

**Comment by saramoein372** [comment]
Sorry Kelvin, for unknown reason my last message got bolded and editing could not fix it.

**Comment by zktuong** [comment]
hi @saramoein372, for `combined_folder`, can you try and see if just specifying `combined_folder = 'tutorial_scgp2'` works?

**Comment by saramoein372** [comment]
Cool! it worked Kelvin. Thank you!

But it does not generates the plots. Is it because of the "Insufficient contigs for running tigger-genotype"?
These are the messages during running:

Processing data file(s) : 100%|██████████| 1/1 [00:01<00:00,  1.06s/it]
      Running tigger-genotype with novel allele discovery.
      Novel allele discovery execution halted.
      Attempting to run tigger-genotype without novel allele discovery.
     Insufficient contigs for running tigger-genotype. Defaulting to original heavy chain v_calls.
            Reconstructing heavy chain dmask germline sequences with v_call.
            Reconstructing light chain dmask germline sequences with v_call.
      For convenience, entries for heavy chain in `v_call` are copied to `v_call_genotyped`.
      For convenience, entries for light chain `v_call` are copied to `v_call_genotyped`.
Writing out to individual folders : 100%|██████████| 1/1 [00:00<00:00,  7.71it/s]
Although tigger-genotype was not run successfully, file will still be saved with `_genotyped.tsv`extension for convenience.

**Comment by zktuong** [comment]
> But it does not generates the plots. Is it because of the "Insufficient contigs for running tigger-genotype"?

That's correct. If `tigger` doesn't manage to work, it will just make a copy of whatever's in the `v_call` column and there will be no further outputs related to `tigger`. You can also manually remove the `v_call_genotyped` column and it shouldn't impact on subsequent analysis.

If this is all, I will close this issue now. Feel free to reopen or open a new issue if required though!

**Comment by saramoein372** [comment]
Sure. These were all my questions about preprocessing step. Thank you.
